### PR TITLE
feat(cli,client): Allow variable expressions in the env file

### LINKED ISSUE #3310: Use dotenv-expand instead of dotenv package to allow variable expressions in the env file
## Problem
Allow env var expansion for schema.prisma
Currently on Heroku, I need to add schema to the DB_URL for the connection to work (e.g. `${DATABASE_URL}?ssl=1&schema=whatever`).
Of course, another solution would be to set `datasources` manually on `PrismaClient`.

## Suggested solution

Minor (non breaking) changes required.

1. Add `dotenv-expand` as a dependency

2. Enable expansion

    ```diff
    import dotenv from 'dotenv'
    + import dotenvExpand from 'dotenv-expand'
    
    - const xxx = dotenv.config(...)
    + const xxx = dotenvExpand(dotenv.config(...))
    ```

    on
    * https://github.com/prisma/prisma/blob/master/src/packages/client/src/runtime/getPrismaClient.ts
    * https://github.com/prisma/prisma/blob/master/src/packages/cli/src/bin.ts

## Alternatives

Current solution (workaround):
```js
const client = new PrismaClient({ datasources: { db: { url: `${process.env.DATABASE_URL!}?ssl=1&schema=schema_name` } } })
```

## Additional context
PR: https://github.com/prisma/prisma/pull/3330

### PR DESCRIPTION
**Pull request by bre7**
closes #3310

### CODE REVIEW

**Comment by Jolg42** [comment]
Hi @bre7 😃 

Thanks for the PR, I'll bring this feature request to the team.

Could you motivate this feature request maybe?

Maybe just a small description of why you want to see it happen inside Prisma? (You could do it outside Prisma maybe?)

**Comment by Jolg42** [comment]
I just saw that you linked an issue 🙈

**Comment by bre7** [comment]
Sure, have a look at #3310. It's mostly to use env vars without the need to configure PrismaClient manually and avoid multiple contact points.

File: src/packages/client/src/runtime/getPrismaClient.ts [hunk 1]
```diff
@@ -387,7 +388,7 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
         '.env',
       )
       if (fs.existsSync(dotEnvPath)) {
-        return parseDotEnv(fs.readFileSync(dotEnvPath, 'utf-8'))
+        return dotenvExpand(parseDotEnv(fs.readFileSync(dotEnvPath, 'utf-8')))
```
Comment (bre7) [inline src/packages/client/src/runtime/getPrismaClient.ts:391]: Needed ? `dotenv-expand` expects config as a parameter, not `parse`'s result https://github.com/motdotla/dotenv-expand/blob/master/lib/main.js#L3

**Comment by matthewmueller** [comment]
Hi @bre7, thanks for the PR!

We're getting quite a few issues around .env. We want to spend some time understanding the problem better before jumping to a solution. We may end up going with this but we want to explore the possible solutions first. 

Thanks for your patience!

**Review by timsuchanek — APPROVED** [review]

**Comment by Akxe** [comment]
I like this, just a few weeks back I was looking at how to do variable in `.env` files to find it is not possible.

I had another library that needed to connect to DB. But it wanted the parts separate. (I ended up using `new URL(dbString)` as the solution but would love to have it separate.

**Comment by timsuchanek** [comment]
@bre7 can you please rebase to master? Then we can merge this :)

**Comment by Jolg42** [comment]
I fixed the conflict, PR needs to be reviewed, specifically this comment https://github.com/prisma/prisma/pull/3330/#discussion_r472607729

**Comment by Jolg42** [comment]
We could add a test so we know this works too

**Comment by bre7** [comment]
missing package.json updates for testing

File: src/packages/client/src/runtime/getPrismaClient.ts (hunk 1 above)
Comment (Jolg42) [reply src/packages/client/src/runtime/getPrismaClient.ts:388]: Indeed I will fix that before merging

**Comment by Jolg42** [comment]
Example usage from a .env file
```
USERNAME="postgres"
DATABASE_URL="postgres://${USERNAME}@localhost:5432/repro"
```

**Comment by Jolg42** [comment]
@bre7 Sorry about that but it looks like you will need to sign the Contributor License Agreement again here https://cla-assistant.io/prisma/prisma?pullRequest=3330

It's because we just updated it with the new company name (content is the same).

**Comment by bre7** [comment]
👍

**Comment by bre7** [comment]
👏 

One last thing: might be worth it to add a comment under https://github.com/prisma/prisma/blob/master/README.md#the-prisma-schema:

```diff
datasource db {
  provider = "postgresql"
-  url      = env("DATABASE_URL")
+  url      = env("DATABASE_URL") /// variable expressions allowed (see dotenv-expand)
}
```

**Comment by bre7** [comment]
Issues in real app. Prisma isn't parsing the schema correctly (assumes public)

**Comment by bre7** [comment]
Missing: add test case with real connection string. Seeing some issues at the moment
> Error validating datasource `db`: The URL for datasource `db` must start with the protocol `postgresql://`.

**Pull request merged**